lib/index.js

```diff
@@ -325,11 +327,35 @@ export function configureUnifiedLanguageServer(
       workspaces.add(event.rootUri)
     }
 
-    if (
+    hasWorkspaceFolderCapability = Boolean(
       event.capabilities.workspace &&
-      event.capabilities.workspace.workspaceFolders
-    ) {
-      connection.workspace.onDidChangeWorkspaceFolders(function (event) {
+        event.capabilities.workspace.workspaceFolders
+    )
+
+    /** @type {InitializeResult} */
+    const result = {
+      capabilities: {
+        textDocumentSync: TextDocumentSyncKind.Full,
+        documentFormattingProvider: true,
+        codeActionProvider: {
+          codeActionKinds: [CodeActionKind.QuickFix],
+          resolveProvider: true
+        }
+      }
+    }
+
+    if (hasWorkspaceFolderCapability) {
+      result.capabilities.workspace = {
+        workspaceFolders: {supported: true, changeNotifications: true}
+      }
+    }
```

Comment:
why mutate like this, instead of returning `result`, and using a ternary to form the `workspace` object based on `hasWorkspaceFolderCapability`?

Not assigning it, and not having to type it, and having the logic *inside* the object seems clearer to me?

Reply:
I literally followed the guide linked in the PR description, but your proposal works too and I agree it looks a bit nicer.